### ISSUE TITLE
Redundant DMA transfer and PIO program make me confused.

### ISSUE THREAD

**Issue by mxyxbb** [dgatf/logic_analyzer_rp2040#2]
Your code really save me a lot of time. Thank you so much.
I have a question about the dma channels used to transfer trigger index, also about the mux pio program.
**Apped: the mux pio program**
```
.program mux
    pull
    mov isr osr
    push
    irq 0
halt:
    jmp halt
```
**Apped: the dma channels used to transfer trigger index**
```
dma_channel_config channel_config_trigger = dma_channel_get_default_config(dma_channel_trigger_[trigger_count_]);
        channel_config_set_transfer_data_size(&channel_config_trigger, DMA_SIZE_32);
        channel_config_set_write_increment(&channel_config_trigger, false);
        channel_config_set_read_increment(&channel_config_trigger, false);
        channel_config_set_dreq(&channel_config_trigger, pio_get_dreq(pio1, sm_trigger_[trigger_count_], false));
        dma_channel_configure(
            dma_channel_trigger_[trigger_count_],
            &channel_config_trigger,
            &pio0->txf[sm_mux_],                       // write address
            &triggered_channel_index_[trigger_count_], // read address
            1,
            true);
```
As far as I am concerned, the trigger sm sends a dma request and rises an interrupt.
The corresponding dma channel transfers the trigger index to the mux sm tx fifo, 
and at the same time, the trigger index is read in the interrupt handler function.
After that, the mux sm sends a dma request to disable pre trigger and enable post trigger.

In my opinion, the procedure can be optimized like this.
the trigger sm sends a dma request and **don't** rise an interrupt.
The corresponding dma channel transfers the trigger index directly to the variable addr previously used in the interrupt.
**Apped: the interrupt handler in your code**
```
static inline void trigger_handler(void)
{
    triggered_channel_ = pio_sm_get_blocking(pio0, sm_mux_);
    pio_interrupt_clear(pio0, 0);
}
```
And then use chain to function link this dma to the dma which is used to disable pre trigger and enable post trigger.
Like `channel_config_set_chain_to(&channel_config_trigger, dma_channel_pio0_ctrl_);`
And change the dreq of dma_channel_pio0_ctrl_ to null.

In summary, the mux pio program is fully redundant, which even may cause sample point missed because of its consuming 3 more sys clocks.  If there's any ignorance in my suggestion?

**Comment by dgatf** [comment]
Thank you for the feedback.

That was indeed my first attempt, but it can raise a race condition which hangs the rp2040.

Though I'll be glad to review any pull request improving the code.

>the trigger sm sends a dma request and rises an interrupt. The corresponding dma channel transfers the trigger index to the mux sm tx fifo,

sm trigger index does not raise an interrupt. It copies the trigger index value to mux fifo. Then mux read the trigger index, raise an interrupt and send the index value. The mux is needed to use its fifo and avoid the race condition. Otherwise you can get a race condition when trying to write to the same memory address by 2 or 3 trigger indexes.
 
Regarding the samples missed. There is no sample missed as sm pre trigger capture is stopped at the same cycle as sm post trigger capture is enabled with _dma_channel_pio0_ctrl__
